### PR TITLE
On-chain Domain Resolution

### PR DESCRIPTION
**Pull request by sunmilee**
On-chain domain resolution of human readable VASP domain names into an on-chain address. Used by DIP-10 Diem ID protocol.

https://github.com/diem/dip/issues/182

### CODE REVIEW

File: dips/dip-182.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+last_updated: 06/015/2021
+issue: https://github.com/diem/dip/issues/182
+---
+
+# Summary
+
+This DIP describes the on-chain domain name resolution process.
+
+On-chain domain name resolution allows organizations to be able to define their product names, and for VASPs to look up other VASPs using the product names.
+Having a human-readable VASP identifier gives the Diem Ecosystem the ability numerate all VASPs by their names, which is useful for merchant purposes as well as p2p transaction.
```
Comment (davidiw) [inline dips/dip-182.md:17]:
```suggestion
Having a human-readable VASP identifier gives the Diem Ecosystem the ability enumerate all VASPs by their names, which is useful for merchant purposes as well as p2p transaction.
```

File: dips/dip-182.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+For example, given a VASP domain name in a payment transaction, VASPs can look up where to route the payment correctly using an approved and accurate mapping of VASP identifier to an on-chain address.
+The Diem blockchain is a natural place to store VASP domain information for domain name resolution. We define here an on-chain domain resolution system that will allow VASPs to query for a VASP's address given a VASP domain.
```
Comment (davidiw) [inline dips/dip-182.md:19]:
```suggestion
```

Comment (davidiw) [reply dips/dip-182.md:19]: kinda repetitive

File: dips/dip-182.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+resource struct DiemIdDomains {
+    domains: vector<DiemIdDomain>,
+}
+struct DiemIdDomain {
+    domain: vector<u8>,  // Utf-8 encoded
+}
+```
+* Field definitions:
+    * `domain` - name of a DiemID domain
+* The `DiemIdDomains` resource can only be published into an account with `Role == PARENT_VASP_ROLE_ID`.
```
Comment (davidiw) [inline dips/dip-182.md:41]: might be worth relabeling these if we push off on the 1.3 release.

Comment (sunmilee) [reply dips/dip-182.md:41]: When is the next cut?

File: dips/dip-182.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+* The `DiemIdDomains` contains a list of `DiemIdDomain` structs that are associated with a VASP. As such, it is possible to register more than one DiemID Domain for a given VASP
+* Only the special TreasuryCompliance account (address `0xB1E55ED`) can manipulate a DiemIdDomains resource:
+    * Only TC account can create and publish a `DiemIdDomains` resource
+    * Only TC account can add, remove or update a `DiemIdDomain` within `DiemIdDomains` resource
+* `DiemIDDomains` resource will be created in an empty state as part of creating a `ParentVASP` account resource, and existing `ParentVASP` accounts without `DiemIDDomains` will have the resource instantiated by the DiemRoot account
+* In order to register a DiemID domain, a VASP needs to submit a request to Diem Association. After approval, Diem Association will use TreasuryCompliance account to issue an on-chain transaction to register a DiemID Domain
```
Comment (aching) [inline dips/dip-182.md:46]: Is the Association responsible for ensuring the uniqueness of all VASP domain names at any given time?  Across all history?  Will domain name re-used be supported, say after a delay, similar to ICANN?

**Review by aching — COMMENTED** [review]
Nice!

File: dips/dip-182.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+This DIP describes the on-chain domain name resolution process.
+
+On-chain domain name resolution allows organizations to be able to define their product names, and for VASPs to look up other VASPs using the product names.
+Having a human-readable VASP identifier gives Diem Ecosystem the ability enumerate all VASPs by their names, which is useful for merchant purposes as well as p2p transaction.
+For example, given a VASP domain name in a payment transaction, VASPs can look up where to route the payment correctly using an approved and accurate mapping of VASP identifier to an on-chain address.
+
+# On-chain data
+
+## Domain Format
+A VASP domain name is a unique string that is mapped to a VASP. Specification:
```
[the remark this answers is not in the document]
Comment (davidiw) [reply dips/dip-182.md:23]: I don't think this is appropriate scope within a DIP.

File: dips/dip-182.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+* Case insensitive
+* Valid regular expression: `^[a-zA-Z0-9][a-zA-Z0-9.]*$`
+* Maximum length: 63 characters (64 including `@`)
```
[the remark this answers is not in the document]
Comment (davidiw) [reply dips/dip-182.md:26]: Haha, legacy cruft 😭

File: dips/dip-182.md (hunk 3 above)
Comment (davidiw) [reply dips/dip-182.md:40]: 1.3 or 1.4 won't matter since the framework release won't happen until 1.4 cut.

File: dips/dip-182.md (hunk 4 above)
Comment (davidiw) [reply dips/dip-182.md:46]: I don't think this is the appropriate doc to describe processes like that.

**Review by davidiw — COMMENTED** [review]
On the one hand, I agree that there are a lot of processes and perhaps it may be worthwhile calling out in an appendix. Though I do have concerns about mixing DA process with technical solutions. Perhaps there should be a secondary DIP on processes if we deem it appropriate.

File: dips/dip-182.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (aching) [reply dips/dip-182.md:23]: I think it's fine to say this is something the Diem Association will provide rules on, however, I think some guidance somewhere is important.  Where will that guidance exist?

File: dips/dip-182.md (hunk 4 above)
Comment (aching) [reply dips/dip-182.md:46]: Where will we have that information?  If folks are to use this feature they will need to know these types of things.

Comment (davidiw) [reply dips/dip-182.md:46]: Fair to some of your points. I agree it is relevant to say that all names are to be unique and that is the responsibility of the TC maintainers to ensure uniqueness.

File: dips/dip-182.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+```
+* Field definitions:
+    * `domain` - name of a VASP domain
+* The `VaspDomains` resource can only be published into an account with `Role == PARENT_VASP_ROLE_ID`
+* The `VaspDomains` contains a list of `VaspDomain` structs that are associated with a VASP. It is possible to register more than one VaspDomain for a given VASP
+* Only the Treasury Compliance (TC) account (address `0xB1E55ED`) can manipulate a `VaspDomains` resource:
+    * Only TC account can create and publish a `VaspDomains` resource
+    * Only TC account can add, remove or update a `VaspDomain` within `VaspDomains` resource
+* `VaspDomains` resource will be created in an empty state as part of creating a `ParentVASP` account resource, and existing `ParentVASP` accounts without `VaspDomains` will have the resource instantiated by the DiemRoot account
+* In order to register a VASP domain, a VASP needs to submit a request to Diem Association. After approval, Diem Association will use TC account to issue an on-chain transaction to register a VaspDomain. Diem Association will ensure uniqueness of domain names across all VASPs.
```
Comment (aching) [inline dips/dip-182.md:46]: If the number of VASPs is not too large, it should be fine for the Diem Association to handle registration, expiration, uniqueness, re-use, and other characteristics.  Before this DIP can move to Accepted, I think we will need to have commitment from the Diem Association that these responsibilities can be taken on.

Comment (sunmilee) [reply dips/dip-182.md:46]: Who do we need to talk to ? @aching @davidiw

File: dips/dip-182.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidiw) [reply dips/dip-182.md:23]: I think there were some hiccups on the guide that are being worked on. There is some content in the operators-runbook repo.

File: dips/dip-182.md (hunk 7 above)
Comment (davidiw) [reply dips/dip-182.md:46]: I think this should be pretty quick conversation with @pdhamdhere .

**Review by davidiw — APPROVED** [review]
/land

**Pull request merged**